ndk/src/font.rs

```diff
@@ -0,0 +1,457 @@
+    pub const EXTRA_LIGHT: FontWeight = FontWeight(ffi::AFONT_WEIGHT_EXTRA_LIGHT as u16);
+    pub const LIGHT: FontWeight = FontWeight(ffi::AFONT_WEIGHT_LIGHT as u16);
+    pub const NORMAL: FontWeight = FontWeight(ffi::AFONT_WEIGHT_NORMAL as u16);
+    pub const MEDIUM: FontWeight = FontWeight(ffi::AFONT_WEIGHT_MEDIUM as u16);
+    pub const SEMI_BOLD: FontWeight = FontWeight(ffi::AFONT_WEIGHT_SEMI_BOLD as u16);
+    pub const BOLD: FontWeight = FontWeight(ffi::AFONT_WEIGHT_BOLD as u16);
+    pub const EXTRA_BOLD: FontWeight = FontWeight(ffi::AFONT_WEIGHT_EXTRA_BOLD as u16);
+    pub const BLACK: FontWeight = FontWeight(ffi::AFONT_WEIGHT_BLACK as u16);
+    pub const MAX: FontWeight = FontWeight(ffi::AFONT_WEIGHT_MAX as u16);
+}
+
+impl fmt::Display for FontWeight {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match *self {
+            FontWeight::THIN => f.write_str("Thin"),
+            FontWeight::EXTRA_LIGHT => f.write_str("Extra Light (Ultra Light)"),
+            FontWeight::LIGHT => f.write_str("Light"),
+            FontWeight::NORMAL => f.write_str("Normal (Regular)"),
+            FontWeight::MEDIUM => f.write_str("Medium"),
+            FontWeight::SEMI_BOLD => f.write_str("Semi Bold (Demi Bold)"),
+            FontWeight::BOLD => f.write_str("Bold"),
+            FontWeight::EXTRA_BOLD => f.write_str("Extra Bold (Ultra Bold)"),
+            FontWeight::BLACK => f.write_str("Black (Heavy)"),
+            _ => writeln!(f, "{}", self.0),
+        }
+    }
+}
+
+/// The error type returned when an invalid font weight value is passed.
+#[derive(Debug)]
+pub struct FontWeightValueError(());
+
+impl fmt::Display for FontWeightValueError {
+    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
+        fmt.write_str("font weight must be positive and less than or equal to 1000")
+    }
+}
+
+impl std::error::Error for FontWeightValueError {}
+
+impl TryFrom<u16> for FontWeight {
+    type Error = FontWeightValueError;
+
+    fn try_from(value: u16) -> Result<Self, Self::Error> {
+        FontWeight::new(value)
+    }
+}
+
+/// A 4-byte integer representing an OpenType axis tag.
+#[derive(Clone, Copy, PartialEq, Eq)]
+pub struct AxisTag(u32);
+
+impl AxisTag {
+    pub const fn from_be(value: u32) -> Result<Self, AxisTagValueError> {
+        // Each byte in a tag must be in the range 0x20 to 0x7E. A space character cannot be
+        // followed by a non-space character. A tag must have one to four non-space characters.
+        // See https://learn.microsoft.com/en-us/typography/opentype/spec/otff#data-types for details.
+
+        let bytes = value.to_be_bytes();
+
+        // Each byte in a tag must be in the range 0x20 to 0x7E.
+        macro_rules! check_byte_range {
+            ($($e:expr)+) => {
+                $(
+                    if !(bytes[$e] as char).is_ascii_graphic() && bytes[$e] != b' ' {
+                        return Err(AxisTagValueError::InvalidCharacter);
```

Comment:
This could also be done with an iterator, instead of a macro?

Reply:
Or... a `for` loop?

Reply:
That implies dropping `const` from the `fn`, I don't see why that is needed.

Reply:
I did write that part with macros to make constructors of `AxisTag` const. I thought `AxisTag::from_be` should be const because it is merely a wrapper of `u32` but with a value validation; since we can't include all kinds of axis tags in `ndk`, some would use `ndk::font` like the following.
```
const WEIGHT_TAG: AxisTag = AxisTag::from_be_bytes(*b"wght");
const MY_CUSTOM_TAG: AxisTag = AxisTag::from_be_bytes(*b"cstm");
```
Without the const constructor, users should use `lazy_static!` for integer constants.

BTW, I should have added constructors that panic for invalid values and return the unwrapped value in const context...

Reply:
Yeah, it seems we can `const`-ly `panic` then. Sure to keep this if you see usefulness in this constructor being const, but a comment above the macro magic explaining this would be nice (for a future contributor to replace once more of `std` becomes `const`-stable).